### PR TITLE
Adds module to configure dns on Cisco UCS Server.

### PR DESCRIPTION
**Pull request by techyragu**
Signed-off-by: Rahul Gupta <ragupta4@cisco.com>

##### SUMMARY

Adds the module to configure dns on Cisco UCS Server.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

cisco_ucs_dns
##### ANSIBLE VERSION

```

```

##### ADDITIONAL INFORMATION

Adds util file `cisco_ucs.py` under `module_utils` to share common utility across all Cisco UCS modules.

```

```

### CODE REVIEW

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (resmo) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: please remove version added in this context. this clause is used to show which param was added in which ansible version.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (resmo) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: please remove version added in this context. this clause is used to show which param was added in which ansible version.

File: lib/ansible/module_utils/cisco_ucs.py
Comment (resmo) [inline lib/ansible/module_utils/cisco_ucs.py]: code under module utils must be released unter bsd-ish license.

**Review by resmo — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Removed "version_added".

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Removed "version_added".

File: lib/ansible/module_utils/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/module_utils/cisco_ucs.py]: License changed to BSD.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (gundalow) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Name needed

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (gundalow) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: `Configures DNS on a Cisco USC server`

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (gundalow) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: As above

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (gundalow) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: `Should the specified DNS server be added or removed.`

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (gundalow) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: `required: false` is the default and can be removed.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (gundalow) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Can you only specify one DNS server?
If this supports multiple change the argspec to `type='list'` and update the code and examples as needed.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Done.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Done.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Done.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Removed `false`

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Module only support single DNS server.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py]: Added `name`.

**Comment by techyragu** [comment]
Is there still anything pending on this PR for approval?

**Review by resmo — APPROVED** [review]

**Comment by techyragu** [comment]
Any update?

File: lib/ansible/module_utils/cisco_ucs.py
Comment (resmo) [inline lib/ansible/module_utils/cisco_ucs.py:73]: IMHO not a blocker but try to avoid catch all exception handling

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (resmo) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:45]: there are some missing docs e.g ucs_* params, these seems to be "generic" and it would make sense to add them to the doc fragments (for inclusion in other ucs modules) See https://github.com/ansible/ansible/tree/devel/lib/ansible/utils/module_docs_fragments

**Review by resmo — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/module_utils/cisco_ucs.py:73]: @resmo - This will require more structural changes at SDK level which is being internally consumed by this module. We will take this as TO DO task to be done in next SDK release and will fix this review comment in another PR later.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:45]: Added common documentation under module_docs_fragments.

**Comment by techyragu** [comment]
@resmo - Could you please verify and approve.

**Review by resmo — APPROVED** [review]

**Comment by resmo** [comment]
The mentioned issues has been addressed, there is catch all exception (which should be avoided) but IMHO not a blocker. 

code LGTM, not tested

@gundalow anything from you?

File: lib/ansible/module_utils/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/module_utils/cisco_ucs.py]: I am surprised this is a function. Where and how will this be used ?

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:102]: I would advice against using these ucs-prefixed names Most (if not all)l existing modules use **hostname**, **username** and **password** as the parameters for authentication. Also **ucs_port**, **ucs_secure** and **ucs_proxy** have standard parameters, the documentation names these **port**, **secure** and **proxy**.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:94]: What is ucs_server used for. The documentation is not very clear about it.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Remove this.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: we tend to use **use_ssl** for this. I know it's a bit outdated, but again, easier for users if we reuse existing parameters.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Don't specify this. It's implicit.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: I would expect a proper default here, probably `default: 'yes'`

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: So we have a default way of doing proxies. Normally it picks up the system settings, unless **use_proxy** is set to 'no'. The format of this parameter is best kept in line with other modules for simplicity.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: It doesn't state in what format.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Remove this. Unclear what the purpose is of this parameter.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
@ragupta-git Could you please get this addressed in the next couple of days?
Cut off for new features (including new modules) is Tuesday 29th

File: lib/ansible/module_utils/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/module_utils/cisco_ucs.py]: Removed this function.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:102]: Prefix `ucs_` is introduced as future cisco ucs modules have parameters as hostname, username, password and port which uses the utils module cisco_ucs. So to avoid conflict we have appended prefix `ucs_`

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: All parameter names are mapped to underlying SDK and APIs i.e. ucsmsdk and ucsm_apis. For maintainabilty purpose we have mapped all the ansible modules to the corresponding ucsm apis.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Here the `default is None`. Hence not defined as `yes or no`.  User can alone mention the port not mandatory to define secure parameter. SDK internally takes care of all the permutation and combination.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Same reason, all parameters being mapped to SDK. Lots of people started using ucsmsdk long back its difficult to change the convention now. It will break the backward compatibility.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Added for future support. This will enable single connection handle can be used across multiple module in a playbook rather making connection handle everytime a module is invoked.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Removed default.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Added the format.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py]: Removed default.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:94]: Added the `Note` under module_fragment. (common documentation)

**Comment by techyragu** [comment]
@gundalow:  I have made the possible changes.
@dagwieers: Please approve as the cut-off date in Tuesday 29th.

**Comment by dagwieers** [comment]
So it seems the packages are stored in a directory of the vendor, whereas we do not tend to do that. We had this discussion as well for **aci** vs **cisco_aci**, **imc** vs **cisco_imc**, **ios** vs **cisco_ios** or **nxos** vs **cisco_nxos**. The decision back then was not to add the vendor in the name.

The same goes for the modules, they are now called **cisco_ucs_dns**, whereas we tend to just call them **ucs_dns.py**.

So instead if **remote_management/cisco_ucs/cisco_ucs_dns.py** it would become **remote_management/ucs/ucs_dns.py**.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:102]: Well, this would be opposite of how other modules are doing it in Ansible. The default **hostname** **username** and **password** are the parameters for the module. The question is, are you doing your users a favor by turning around the logic ?

File: lib/ansible/module_utils/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/module_utils/cisco_ucs.py:58]: I think this is confusing. Why not use `self.module.params` everywhere instead ? At least that's much more clear when reading the code than a localy define variable named `ansible`.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:29]: No trailing dot in the short_description. Keep the index consistent.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:36]: Start description with capital, trailing dot.
The description is not correct, "present" does not always add a DNS, it makes sure it is define. Otherwise it would not be idempotent.

Comment (dagwieers) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:36]: Please do this for all descriptions.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:38]: Quotes are not needed for strings in YAML, so we don't use them unless required.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:47]: Don't quote strings here.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:55]: The author entry should be a list, and it should list individuals with their github account.
Otherwise you will not be notified for issues/PRs.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py:49]: There is an inconsistency here, I actually prefer this over the implemented **ucs_port**.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py:53]: The standard parameter for this is **use_ssl*, please change this.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py:56]: There is no default here ?

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py:57]: So by default Ansible modules use the system-configured proxy settings, unless you set **use_proxy** to **no**.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/utils/module_docs_fragments/cisco_ucs.py:36]: Not consistent with other modules **hostname**, but also not consistent with the implementation.

**Comment by dagwieers** [comment]
@ragupta-git I am not sure if this module is ready for v2.4 yet. I send you a mail at your Cisco address wrt. this.
- I am not convinced the module interface is conform other modules (both the module name, location as well as parameters)
- If the interface may still change, it's better not to set it in stone right now, and wait for Ansible v2.5
- It's only the DNS module, I expect to see maybe 30 or more UCS modules, so maybe it's better to wait until you have a collection that is approved, rather than push this one now ?

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/module_utils/cisco_ucs.py:47]: Why not just do:
```python
        if not HAS_UCSMSDK:
            fail_json(msg='ucsmsdk is required for this module')
        if not HAS_UCSMAPI:
            fail_json(msg='ucsm_apis is required for this module')
```

File: lib/ansible/module_utils/cisco_ucs.py
Comment (dagwieers) [inline lib/ansible/module_utils/cisco_ucs.py:80]: So I would do this differently. We just designed a similar interface for Cisco ACI and to make the modules as minimal as possible, you can simply do the login automatically when you create the object. And this would ensure this object would already exist. within the object.

Also I don't think **ucs_server** is a good name for the purpose. Why not call it **login_handle** or something similar ?

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:76]: I am not sure this is the best interface for every module.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:97]: This should probably go to a common argument_spec in the common module_utils library.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by techyragu** [comment]
@dagwieers : I understand `cisco_ucs_dns` can be changed to `ucs_dns`. But why folder name  to be changed from`cisco` to `ucs`. I can see other vendors also have their specific folder let say `google or amazon or docker`.  Also there is already a folder named `imc` which is not added by us and we have similar modules coming for Cisco IMC too, hence we can have `cisco` folder and all our UCS and IMC modules within it.

**Comment by gundalow** [comment]
Understand `modules/network/` we use the product (or product OS) name
Eg
`network/ios` for Cisco IOS
`network/iosxr` for Cisco IOS XR

**Comment by dagwieers** [comment]
@ragupta-git There is another reason for separating. Using *.github/BOTMETA.yml* you can give different maintainers access to different directories so that new submissions end up to different maintainers.

anazon is a good example, I think there's good cause to separating modules. docker is also the name of the technology. For vmware we may split the vca and vcenter stuff as well at some point.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py:57]: By default `proxy=None`.  This implies module will use environment proxy.
If any value is assigned to `proxy` parameter , module will use the respective value.
Note: `proxy` is not bool parameter.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py:56]: As mentioned in earlier comments ```Here the default is None. Hence not defined as yes or no. User can alone mention the port or secure option, it is not mandatory to define secure parameter. SDK internally takes care of all the permutation and combination.```

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:29]: Done

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:102]: @dagwieers: I understand its bit off from the standard names used by other ansible modules. However our modules is mapped to the inherent UCS schema so in longer run it will be useful for users too as the future modules will have parameters with names `hostname, username, password`. So as to avoid the clash in the future we opted for this `ucs_` convention.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:36]: Do we need to really explicitly mention the idempotent behaviour in the definition. Ansible module behaviour by default is `idempotent`.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:38]: Removed quotes.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (techyragu) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:47]: Removed quotes.

File: lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py
Comment (dagwieers) [reply lib/ansible/modules/remote_management/cisco/cisco_ucs_dns.py:36]: The current description is wrong, the module doesn't add the DNS server.
And no, not all modules are idempotent, so we cannot assume that the user can expect idempotent behavior. The description is easy to fix.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py:56]: I'd rather have the standard behavior from other modules, which is to specify **use_ssl**. Check **uri** module.

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (dagwieers) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py:57]: No, obviously **proxy** is not a boolean, but **use_proxy** is.

So the standard behavior of Ansible modules is to use the system's configuration (**http_proxy**, **https_proxy**) but in order to disable this default behavior, there is the **use_proxy** option, that defaults to **yes**. And if you want to use a different proxy, you can set the **proxy** parameter.

In you case there is no way to tell the module *not* to use the system-configured proxy server.

File: lib/ansible/module_utils/cisco_ucs.py
Comment (dagwieers) [reply lib/ansible/module_utils/cisco_ucs.py:80]: Other modules using a REST API have a parameter **token** which consists of a token (usually a hash or serialized input) needed to use an existing login.

Your settings for UCS can easily be added using YAML anchors and references. See: https://en.wikipedia.org/wiki/YAML

```yaml
- hosts: localhost

  vars:
    ucs_parameters: &ucs_server
      hostname: 1.2.3.4
      username: admin
      password: c1sco!123
      no_proxy: yes

  tasks:
  - ucs_dns:
      <<: *ucs_server
      name: 1.2.3.4
      description: Internal DNS server

  - ucs_hostname:
      <<: *ucs_server
      name: foobar.ucs.cisco.com
      description: Internal server name
```

File: lib/ansible/utils/module_docs_fragments/cisco_ucs.py
Comment (techyragu) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py:57]: `proxy=None` - Uses system defined Proxy
`proxy={}` - Not to use system defined Proxy.
`proxy=http:\\xx.xx.xx.xx:port` - Use this specific Proxy

Comment (dagwieers) [reply lib/ansible/utils/module_docs_fragments/cisco_ucs.py:57]: @ragupta-git So I feel I have to repeat myself over and over again. Please use the **use_proxy** variable for disabling proxy support. This is how most (if not all) other modules are doing it.

**Comment by dagwieers** [comment]
@dsoper2 @movinalot What is the plan for this older UCS module ?

**Comment by dsoper2** [comment]
@dagwieers We don't plan to pursue this module so this PR can be closed.

**Pull request closed without merging**
